src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -0,0 +1,1531 @@
+        {
+            switch (member)
+            {
+                case PSMemberInfo psMemberInfo:
+                    return psMemberInfo.IsHidden;
+                case MemberInfo memberInfo:
+                    return memberInfo.GetCustomAttributes(typeof(HiddenAttribute), false).Any();
+                case PropertyMemberAst propertyMember:
+                    return propertyMember.IsHidden;
+                case FunctionMemberAst functionMember:
+                    return functionMember.IsHidden;
+            }
+
+            return false;
+        }
+
+        private static bool IsConstructor(object member)
+        {
+            var psMethod = member as PSMethod;
+            var methodCacheEntry = psMethod?.adapterData as DotNetAdapter.MethodCacheEntry;
+            return methodCacheEntry != null && methodCacheEntry.methodInformationStructures[0].method.IsConstructor;
+        }
+    }
+
+    internal class TypeInferenceVisitor : ICustomAstVisitor2
+    {
+        private readonly TypeInferenceContext _context;
+
+
+        public TypeInferenceVisitor(TypeInferenceContext context)
+        {
+            _context = context;
+        }
+
+        private IEnumerable<PSTypeName> InferTypes(Ast ast)
+        {
+            return _context.InferType(ast, this);
+        }
+
+        object ICustomAstVisitor.VisitTypeExpression(TypeExpressionAst typeExpressionAst)
+        {
+            return new[] { new PSTypeName(typeExpressionAst.StaticType) };
+        }
+
+        object ICustomAstVisitor.VisitMemberExpression(MemberExpressionAst memberExpressionAst)
+        {
+            return InferTypesFrom(memberExpressionAst);
+        }
+
+        object ICustomAstVisitor.VisitInvokeMemberExpression(InvokeMemberExpressionAst invokeMemberExpressionAst)
+        {
+            return InferTypesFrom(invokeMemberExpressionAst);
+        }
+
+        object ICustomAstVisitor.VisitArrayExpression(ArrayExpressionAst arrayExpressionAst)
+        {
+            return new[] { new PSTypeName(typeof(object[])) };
+        }
+
+        object ICustomAstVisitor.VisitArrayLiteral(ArrayLiteralAst arrayLiteralAst)
+        {
+            return new[] { new PSTypeName(typeof(object[])) };
+        }
+
+        object ICustomAstVisitor.VisitHashtable(HashtableAst hashtableAst)
+        {
+            return new[] { new PSTypeName(typeof(Hashtable)) };
+        }
+
+        object ICustomAstVisitor.VisitScriptBlockExpression(ScriptBlockExpressionAst scriptBlockExpressionAst)
+        {
+            return new[] { new PSTypeName(typeof(ScriptBlock)) };
+        }
+
+        object ICustomAstVisitor.VisitParenExpression(ParenExpressionAst parenExpressionAst)
+        {
+            return parenExpressionAst.Pipeline.Accept(this);
+        }
+
+        object ICustomAstVisitor.VisitExpandableStringExpression(ExpandableStringExpressionAst expandableStringExpressionAst)
+        {
+            return new[] { new PSTypeName(typeof(string)) };
+        }
```

Comment:
When the type is a constant like this, we should probably return a static to minimize garbage.